src/Persistence_matrix/include/gudhi/matrix.h

```diff
@@ -0,0 +1,1588 @@
+
+  //*** swap/vine
+  // base
+  // boundary: does not update barcode
+  void swap_columns(index columnIndex1, index columnIndex2);
+  // base
+  // boundary: does not update barcode
+  void swap_rows(index rowIndex1, index rowIndex2);
+  // ru: returns true if barcode was changed
+  // pos to id
+  bool vine_swap_with_z_eq_1_case(pos_index index);  // by column position with ordered column container
+  // chain: returns index which was not modified, ie new i+1
+  // id to pos
+  index vine_swap_with_z_eq_1_case(index columnIndex1,
+                                   index columnIndex2);  // by column id with potentielly unordered column container
+  // ru: returns true if barcode was changed
+  // pos to id
+  bool vine_swap(pos_index index);  // by column position with ordered column container
+  // chain: returns index which was not modified, ie new i+1
+  // id to pos
+  index vine_swap(index columnIndex1, index columnIndex2);  // by column id with potentielly unordered column container
+
+  //*** Representative cycles
+  // chain
+  // ru
+  // id to pos
+  // pos to id
+  void update_representative_cycles();
+  // chain
+  // ru
+  // id to pos
+  // pos to id
+  const std::vector<cycle_type>& get_representative_cycles();
+  // chain
+  // ru
+  // id to pos
+  // pos to id
+  const cycle_type& get_representative_cycle(const Bar& bar);
+
+ private:
+  using matrix_type = 
+    typename std::conditional<
+        isNonBasic,
+        typename std::conditional<
+            Options::is_of_boundary_type,
+            typename std::conditional<
+                Options::has_vine_update || Options::can_retrieve_representative_cycles,
+                typename std::conditional<
+                    Options::column_indexation_type == Column_indexation_types::CONTAINER || 
+                      Options::column_indexation_type == Column_indexation_types::POSITION,
+                    RU_matrix_type, Id_to_index_overlay<RU_matrix_type, Matrix<Options> >
+                >::type,
+                typename std::conditional<
+                    Options::column_indexation_type == Column_indexation_types::CONTAINER || 
+                      Options::column_indexation_type == Column_indexation_types::POSITION,
+                    Boundary_matrix_type,
+                    Id_to_index_overlay<Boundary_matrix_type, Matrix<Options> >
+                >::type
+            >::type,
+            typename std::conditional<
+                Options::column_indexation_type == Column_indexation_types::CONTAINER, 
+                Chain_matrix_type,
+                typename std::conditional<
+                    Options::column_indexation_type == Column_indexation_types::POSITION,
+                    Position_to_index_overlay<Chain_matrix_type, Matrix<Options> >,
+                    Id_to_index_overlay<Chain_matrix_type, Matrix<Options> >
+                >::type
+            >::type
+        >::type,
+        Base_matrix_type
+    >::type;
```

Comment:
I don't know if this would be more digest with `if constexpr`. In summary:

```mermaid
graph LR
    A((Matrix_type)) --> B{has_column_pairings OR\n has_vine_update OR\n can_retrieve_representative_cycles}
    B -- false --> C(Base_matrix_type)
    B -- true --> D{is_of_boundary_type}
    D -- false --> E(Chain_matrix_type)
    D -- true --> F{has_vine_update OR\n can_retrieve_representative_cycles}
    F -- true --> G(RU_matrix_type)
    F -- false --> H(Boundary_matrix_type)
```

Reply:
Forget my point, you are right, I was thinking how to write it with a lighter syntax, but it is not working...

Reply:
Using `std::conditional_t` would allow to remove some `typename` and `::type`. Not much I know...